Source/WebKit/Shared/Extensions/WebExtensionRegisteredScriptsSQLiteStore.cpp

```diff
@@ -0,0 +1,303 @@
+            completionHandler(errorMessage);
+            return;
+        }
+
+        protectedThis->addScripts(scripts, [completionHandler = WTFMove(completionHandler)](const String& errorMessage) mutable {
+            completionHandler(errorMessage);
+        });
+    });
+}
+
+void WebExtensionRegisteredScriptsSQLiteStore::deleteScriptsWithIDs(Vector<String> ids, CompletionHandler<void(const String& errorMessage)>&& completionHandler)
+{
+    if (ids.isEmpty()) {
+        completionHandler({ });
+        return;
+    }
+
+    queue().dispatch([weakThis = ThreadSafeWeakPtr { *this }, ids = crossThreadCopy(ids), completionHandler = WTFMove(completionHandler)]() mutable {
+        RefPtr protectedThis = weakThis.get();
+        if (!protectedThis) {
+            completionHandler({ });
+            return;
+        }
+
+        String errorMessage;
+        if (!protectedThis->openDatabaseIfNecessary(errorMessage, false)) {
+            WorkQueue::mainSingleton().dispatch([errorMessage = crossThreadCopy(errorMessage), completionHandler = WTFMove(completionHandler)]() mutable {
+                completionHandler(errorMessage);
+            });
+            return;
+        }
+
+        ASSERT(errorMessage.isEmpty());
+
+        DatabaseResult result = SQLiteDatabaseExecute(*(protectedThis->database()), makeString("DELETE FROM registered_scripts WHERE key in ("_s, rowFilterStringFromRowKeys(ids), ")"_s));
+        if (result != SQLITE_DONE) {
+            RELEASE_LOG_ERROR(Extensions, "Failed to delete scripts for extension %s.", protectedThis->uniqueIdentifier().utf8().data());
+            errorMessage = "Failed to delete scripts from registered content scripts storage."_s;
+        }
+
+        String deleteDatabaseErrorMessage = protectedThis->deleteDatabaseIfEmpty();
+
+        WorkQueue::mainSingleton().dispatch([deleteDatabaseErrorMessage = crossThreadCopy(deleteDatabaseErrorMessage), errorMessage = crossThreadCopy(errorMessage), completionHandler = WTFMove(completionHandler)]() mutable {
+            // Errors from opening the database or deleting keys take precedence over an error deleting the database.
+            completionHandler(!errorMessage.isEmpty() ? errorMessage : deleteDatabaseErrorMessage);
+        });
+    });
+}
+
+void WebExtensionRegisteredScriptsSQLiteStore::addScripts(Vector<Ref<JSON::Object>> scripts, CompletionHandler<void(const String& errorMessage)>&& completionHandler)
+{
+    // Only save persistent scripts to storage
+    Vector<Ref<JSON::Object>> persistentScripts;
+    for (const auto& script : scripts) {
+        if (auto persistent = script->getBoolean(persistAcrossSessionsKey); persistent.value())
+            persistentScripts.append(script);
+    }
+
+    if (persistentScripts.isEmpty()) {
+        completionHandler({ });
+        return;
+    }
+
+    queue().dispatchSync([weakThis = ThreadSafeWeakPtr { *this }, persistentScripts, completionHandler = WTFMove(completionHandler)]() mutable {
```

Comment:
This should not be `dispatchSync()`.

```suggestion
    queue().dispatch([weakThis = ThreadSafeWeakPtr { *this }, persistentScripts, completionHandler = WTFMove(completionHandler)]() mutable {
```